src/lib.rs

```diff
@@ -1,78 +1,94 @@
 extern crate rand;
 
 #[cfg(feature = "serde")]
 #[macro_use]
 extern crate serde;
 
-use num::{cast::FromPrimitive, ToPrimitive, Unsigned};
-use std::{cmp::Ord, marker::PhantomData};
-
-const WORD_SIZE: usize = 64;
+use num_traits::{cast::FromPrimitive, AsPrimitive, PrimInt, ToPrimitive, Unsigned};
+use std::{cmp::Ord, marker::PhantomData, mem::size_of};
 
 #[derive(Debug)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
-pub struct PackedVec<T> {
+pub struct PackedVec<T, StorageT = usize> {
     len: usize,
-    bits: Vec<u64>,
+    bits: Vec<StorageT>,
     item_width: usize,
     phantom: PhantomData<T>,
 }
 
-impl<'a, T> PackedVec<T>
+impl<'a, T> PackedVec<T, usize>
 where
-    T: FromPrimitive + Ord + ToPrimitive + Unsigned,
+    T: 'static + AsPrimitive<usize> + FromPrimitive + Ord + PrimInt + ToPrimitive + Unsigned,
+    usize: AsPrimitive<T>,
 {
-    /// Return a `PackedVec` containing a compressed version of the elements of
-    /// `vec`.
-    pub fn new(vec: Vec<T>) -> PackedVec<T> {
-        let len = vec.len();
-        let item_width = if let Some(v) = vec.iter().max() {
-            i_log2((*v).to_u64().unwrap())
-        } else {
+    /// Constructs a new `PackedVec` from `vec`. The `PackedVec` has `usize` backing storage, which
+    /// is likely to be the best choice in nearly all situations.
+    pub fn new(vec: Vec<T>) -> Self {
+        PackedVec::new_with_storaget(vec)
+    }
+}
+
+impl<'a, T, StorageT> PackedVec<T, StorageT>
+where
+    T: 'static + AsPrimitive<StorageT> + FromPrimitive + Ord + PrimInt + ToPrimitive + Unsigned,
+    StorageT: AsPrimitive<T> + PrimInt + Unsigned,
+{
+    /// Constructs a new `PackedVec` from `vec` (with a user-defined backing storage type).
+    pub fn new_with_storaget(vec: Vec<T>) -> PackedVec<T, StorageT> {
+        if size_of::<T>() > size_of::<StorageT>() {
+            panic!("The backing storage type must be the same size or bigger as the stored integer size.");
+        }
+        let m = vec.iter().max();
+        // If the input vector was empty, or if it consisted entirely of zeros, we don't need to
+        // fill the backing storage with anything.
+        if m.is_none() || *m.unwrap() == T::zero() {
             return PackedVec {
-                len: 0,
+                len: vec.len(),
                 bits: vec![],
                 item_width: 0,
                 phantom: PhantomData,
             };
         };
+
+        let item_width = i_log2(*m.unwrap());
         let mut bit_vec = vec![];
-        let mut buf: u64 = 0;
+        let mut buf = StorageT::zero();
         let mut bit_count: usize = 0;
-        for ref item in vec.iter() {
-            let item = item.to_u64().unwrap();
-            if bit_count + item_width < WORD_SIZE {
-                let shifted_item = item << WORD_SIZE - (item_width + bit_count);
-                buf |= shifted_item;
+        for item in &vec {
+            let item: StorageT = (*item).as_();
```

Comment:
What does `as_` do?

Reply:
It's basically the `as` keyword in trait form https://docs.rs/num-traits/0.2.6/num_traits/cast/trait.AsPrimitive.html#tymethod.as_